include/net/net_if.h

```diff
@@ -664,6 +664,70 @@ struct net_if_mcast_addr *net_if_ipv6_maddr_lookup(const struct in6_addr *addr,
 						   struct net_if **iface);
 
 /**
+ * @typedef net_if_mcast_callback_t
+
+ * @brief Define callback that is called whenever IPv6 multicast address group
+ * is joined or left.
+
+ * @param "struct net_if *iface" A pointer to a struct net_if to which the
+ *        multicast address is attached.
+ * @param "const struct in6_addr *addr" IPv6 multicast address.
+ * @param "bool is_joined" Is the address joined or not.
+ */
+typedef void (*net_if_mcast_callback_t)(struct net_if *iface,
+					const struct in6_addr *addr,
+					bool is_joined);
+
+/**
+ * @brief Multicast monitor handler struct.
+ *
+ * Stores the multicast callback information. Caller must make sure that
+ * the variable pointed by this is valid during the lifetime of
+ * registration. Typically this means that the variable cannot be
+ * allocated from stack.
+ */
+struct net_if_mcast_monitor {
+	/** Node information for the slist. */
+	sys_snode_t node;
+
+	/** Network interface */
+	struct net_if *iface;
+
+	/** Multicast callback */
+	net_if_mcast_callback_t cb;
+};
+
+#if defined(CONFIG_NET_IPV6_MLD)
+/**
+ * @brief Register a multicast monitor
+ *
+ * @param mon Monitor handle
```

Comment:
I'd call it "pointer to a monitor storage structure (should be allocated, but doesn't need to be initialized)".